server/src/main/java/org/elasticsearch/action/admin/cluster/health/TransportClusterHealthAction.java

```diff
@@ -236,7 +246,7 @@ private void executeHealth(
             final ClusterStateObserver.Listener stateListener = new ClusterStateObserver.Listener() {
                 @Override
                 public void onNewClusterState(ClusterState newState) {
-                    onNewClusterStateAfterDelay.accept(newState);
+                    executor.execute(() -> onNewClusterStateAfterDelay.accept(newState));
```

Comment:
Hmm in two of the call sites `onNewClusterStateAfterDelay` is lightweight (it calls `waitForEventsAndExecuteHealth` which just submits a task to the master service) so forking here shouldn't be necessary. It's only the one call in `masterOperation` which needs to fork.